agent/util-scripts/pbench-tool-meister-start

```diff
@@ -732,13 +734,89 @@ def main(_prog: str, cli_params: Namespace) -> int:
+        origin_ip = set()
+        any_remote = False
+        template = TemplateSsh(ssh_cmd, shlex.split(ssh_opts), "echo ${SSH_CONNECTION}")
+        for host in tool_group.hostnames.keys():
+            if not localhost.is_local(host):
+                any_remote = True
+                template.start(host)
+
+        for host, params in tool_group.hostnames.items():
+            if not localhost.is_local(host):
+                connection = template.wait(host)
+                logger.debug("Host %s reports connection `%s`", host, connection)
+                if connection.status == 0 and connection.stdout:
+                    # The SSH_CONNECTION value is the full `stdout` and we
+                    # don't expect a stderr on success; the format is
+                    #   "origin_node origin_port local_node local_port"
+                    # we only need the origin node because we know that's an
+                    # IP for our local TDS and Redis host that the remote can
+                    # reach.
+                    origin = connection.stdout.split()[0]
+                    origin_ip.add(origin)
+
+                    # Store the origin reported by this remote in its parameter
+                    # block: we'll program this as the connection address when
+                    # we send the init handshake to that remote later.
+                    params["origin_host"] = origin
+                else:
+                    # If `ssh` fails, we can't orchestrate anything on this
+                    # remote, so terminate with an error.
+                    logger.error("Host %s reports %s", host, connection)
+                    return ReturnCode.REMOTENOTREACHABLE
+
+        # Process the collected origin addresses from our remotes. Ideally we
+        # have only a single entry here, since the current BaseServer init
+        # (and the Bottle WSGI server) doesn't support listening on multiple
+        # addresses; however we're going to just take the first returned origin
+        # address and hope they can all connect.
+        #
+        # NOTE: an alternative would be (in the absence of an explicit
+        # connection spec) to always listen on '0.0.0.0', but this eliminates
+        # IPv6, and we currently lack the mechanism to listen on both '0.0.0.0'
+        # and '::' (the IPv6 equivalent, aka '0:0:0:0:0:0:0:0').
+        #
+        # NOTE: If the caller/human supplied explicit PBENCH_REDIS_SERVER or
+        # PBENCH_TOOL_DATA_SINK (--redis-server or --tool-data-sink), we'll use
+        # those, but by default we'll use the ssh connection origin instead of
+        # the local `hostname` which may not be reachable by the remotes (or
+        # necessarily routable at all).
+        if len(origin_ip) != 1 and any_remote:
+            logger.warning(
+                "Remote hosts don't report a single controller "
+                "origin IP: this may be a problem: origins are %s",
+                origin_ip
```

Comment:
I suggest replacing line 807 with something like
```
                origin_ip if origin_ip else "<none>"
```
(I really don't like the `origins are set()` output.)

Reply:
These are log messages, and it seems silly to waste the effort to make them more "pretty" than necessary to convey the debugging information. However, you have a point in that here, as in much of the agent code, we seem to be using Python's logger to generate end-user messages. I really don't like that, but I suppose I'm stuck with it here. 😬

Reply:
> it seems silly to waste the effort to make them more "pretty" than necessary

I agree; however, given that I might be the poor slob who has to read them, future-me would likely appreciate it if past-me (or past-you) made them "pretty enough".  (Without a certain amount of context, I think that `set()` is likely to be confusing, whereas `{}` or `<none>` is pretty clear.)